Automated cherry pick of #105273: tests: Wait for pod collection to enter a Running state

claudiubelu opened the pull request:
Cherry pick of #105273 on release-1.22.

#105273: tests: Wait for pod collection to enter a Running state


#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind flake
/sig testing
/sig windows

/priority important-soon

#### What this PR does / why we need it:

While running tests in parallel, especially those with higher loads than others, it might take some time for Pods to be Running, even more so if the image has to be pulled as well.

The test ``[sig-node] Pods should delete a collection of pods [Conformance]`` only waits for the for the pods to be scheduled before deleting them, and expects them to be gone in 1 minute, which can flake because of the above reasons. Note that the operations are in order, and kubelet runs them in order, which means that the pod first has to enter the Running state before attempting to delete it.

This commit waits for the Pods to enter the Running state first before deleting the entire collection.

Co-Authored-By: Antonio Ojea <aojea@redhat.com>

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #103671

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
NONE
```

claudiubelu committed:
tests: Wait for pod collection to enter a Running state

While running tests in parallel, especially those with higher loads
than others, it might take some time for Pods to be Running, even more
so if the image has to be pulled as well.

The test [sig-node] Pods should delete a collection of pods [Conformance]
only waits for the for the pods to be scheduled before deleting them, and
expects them to be gone in 1 minute, which can flake because of the above
reasons. Note that the operations are in order, and kubelet runs them in
order, which means that the pod first has to enter the Running state
before attempting to delete it.

This commit waits for the Pods to enter the Running state first before
deleting the entire collection.

Co-Authored-By: Antonio Ojea <aojea@redhat.com>

claudiubelu commented:
/kind flake
/priority important-soon

/sig windows
/sig testing

claudiubelu commented:
/test pull-kubernetes-typecheck

SergeyKanzhelev commented:
I see a flake. Looks ok to merge. Details here: https://github.com/kubernetes/kubernetes/pull/106184#issuecomment-966451199

/approve
/lgtm
/triage accepted

claudiubelu commented:
/assign @tallclair

claudiubelu commented:
/assign @Random-Liu

jsturtevant commented:
/easycla

tallclair commented on test/e2e/common/node/pods.go:
```diff
@@ -866,6 +866,8 @@ var _ = SIGDescribe("Pods", func() {
 				}}, metav1.CreateOptions{})
 			framework.ExpectNoError(err, "failed to create pod")
 			framework.Logf("created %v", podTestName)
+			framework.ExpectNoError(e2epod.WaitForPodNameRunningInNamespace(f.ClientSet, podTestName, f.Namespace.Name))
```
Better to use `e2epod.WaitForsRunningReady` (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go#L107) to wait for all 3 pods in parallel.

claudiubelu replied:
Done

tallclair commented on test/e2e/common/node/pods.go:
```diff
@@ -866,6 +866,8 @@ var _ = SIGDescribe("Pods", func() {
+			framework.Logf("running and ready %v", podTestName)
 		}
 
 		// wait as required for all 3 pods to be found
```
If you use `WaitForPodsRunningReady` as suggested above, then you can delete this check.

claudiubelu replied:
Done

tallclair commented:
Whoops, missed that this is a cherry-pick. I still think waiting for the pods to be ready in parallel is better, rather than serially creating & waiting for each pod. It will cut down on the overall test time, and probably be less flaky (in case something happens to the first pod by the time the 3rd is finished)

marosset commented:
@claudiubelu PTAL at Tim's feedback

claudiubelu committed:
tests: Updates the should delete a collection of pods test

We should wait for all 3 pods after we've spawned them, instead of
spawning and waiting for them sequentially.

claudiubelu commented:
> Whoops, missed that this is a cherry-pick. I still think waiting for the pods to be ready in parallel is better, rather than serially creating & waiting for each pod. It will cut down on the overall test time, and probably be less flaky (in case something happens to the first pod by the time the 3rd is finished)

Done: https://github.com/kubernetes/kubernetes/pull/108593

claudiubelu commented:
/test pull-kubernetes-e2e-kind-ipv6

marosset commented:
/lgtm

jsturtevant commented:
/assign @tallclair

tallclair commented:
Looks like these changes are covered by https://github.com/kubernetes/kubernetes/pull/109789, should we just merge that instead?

claudiubelu commented:
> Looks like these changes are covered by #109789, should we just merge that instead?

The PR you're mentioning is a cherry-pick for v1.23, while this one is for v1.22.

The PR you mentioned only cherry-pick https://github.com/kubernetes/kubernetes/pull/108593 onto v1.23 since the original fix (https://github.com/kubernetes/kubernetes/pull/105273) was already cherry-picked on v1.23 and merged.

This PR cherry-picks both https://github.com/kubernetes/kubernetes/pull/105273 and https://github.com/kubernetes/kubernetes/pull/108593 on v1.22, hence why this PR has 2 commits. The second commit came as a suggestion from you, see above.

claudiubelu commented:
/asign kubernetes/release-managers

claudiubelu commented:
/assign kubernetes/release-managers

claudiubelu commented:
/cc kubernetes/release-managers

Verolop commented:
/lgtm

marosset commented:
/retest
